src/os/bluestore/BlueStore.cc

```diff
@@ -1170,41 +1173,42 @@ struct LruOnodeCacheShard : public BlueStore::OnodeCacheShard {
     ceph_assert(p != lru.begin());
     --p;
     ceph_assert(num >= n);
-    num -= n;
     while (n-- > 0) {
       BlueStore::Onode *o = &*p;
       dout(20) << __func__ << "  rm " << o->oid << " "
-               << o->nref << " " << o->cached << " " << o->pinned << dendl;
+               << o->nref << " " << o->cached << dendl;
+
       if (p != lru.begin()) {
         lru.erase(p--);
       } else {
         ceph_assert(n == 0);
         lru.erase(p);
       }
       *(o->cache_age_bin) -= 1;
-      auto pinned = !o->pop_cache();
-      ceph_assert(!pinned);
-      o->c->onode_map._remove(o->oid);
+      if (o->pin_nref > 1) {
+        dout(20) << __func__ << " " << this << " " << " " << " " << o->oid << dendl;
+      } else {
+	ceph_assert(num);
+        --num;
+        o->clear_cached();
+        o->c->onode_space._remove(o->oid);
+      }
     }
   }
-  void move_pinned(OnodeCacheShard *to, BlueStore::Onode *o) override
+  void _move_pinned(OnodeCacheShard *to, BlueStore::Onode *o) override
   {
     if (to == this) {
       return;
     }
-    ceph_assert(o->cached);
-    ceph_assert(o->pinned);
-    ceph_assert(num);
-    ceph_assert(num_pinned);
-    --num_pinned;
-    --num;
-    ++to->num_pinned;
-    ++to->num;
+    _rm(o);
+    ceph_assert(o->nref > 1);
+    to->_add(o, 0);
   }
   void add_stats(uint64_t *onodes, uint64_t *pinned_onodes) override
   {
+    std::lock_guard l(lock);
     *onodes += num;
-    *pinned_onodes += num_pinned;
+    *pinned_onodes += num - lru.size();
```

Comment:
This might not be accurate anymore, since we are doing a lazy pinning now.

Reply:
That's true if we name onode pinned when it has got nref > 1. If we name onode pinned when it's not in cache LRU - it would be fine ;)
Not a big deal IMO. The primary reason behind the stat is just to make sure that "pinned" onodes aren't leaking or something.. There is no specific need to do that with absolute accuracy...

Reply:
Agree :)